src/Storages/ExternalDataSourceConfiguration.h

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <Parsers/ASTIdentifier.h>
+#include <Interpreters/Context.h>
+#include <Poco/Util/AbstractConfiguration.h>
+
+
+namespace DB
+{
+
+struct ExternalDataSourceConfiguration
+{
+    String host;
+    UInt16 port = 0;
+    String username;
+    String password;
+    String database;
+    String table;
+    String schema;
+
+    std::vector<std::pair<String, UInt16>> addresses; /// Failover replicas.
+
+    String toString() const;
+
+    void set(const ExternalDataSourceConfiguration & conf);
+};
+
+using ExternalDataSourceConfigurationPtr = std::shared_ptr<ExternalDataSourceConfiguration>;
+
+
+struct StoragePostgreSQLConfiguration : ExternalDataSourceConfiguration
+{
+    String on_conflict;
+};
+
+
+struct StorageMySQLConfiguration : ExternalDataSourceConfiguration
+{
+    bool replace_query = false;
+    String on_duplicate_clause;
+};
+
+struct StorageMongoDBConfiguration : ExternalDataSourceConfiguration
+{
+    String collection;
+    String options;
+};
+
+
+using EngineArgs = std::vector<std::pair<String, DB::Field>>;
+
+/* If storage engine's configuration was define via named_collections,
+ * return all options in ExternalDataSource::Configuration struct.
+ *
+ * Also check if engine arguments have key-value defined configuration options:
+ * ENGINE = PostgreSQL(postgresql_configuration, database = 'postgres_database');
+ * In this case they will override values defined in config.
+ *
+ * If there are key-value arguments apart from common: `host`, `port`, `username`, `password`, `database`,
+ * i.e. storage-specific arguments, then return them back in a set: ExternalDataSource::EngineArgs.
+ */
+std::optional<std::tuple<ExternalDataSourceConfiguration, EngineArgs>>
```

Comment:
Please, get rid of `std::tuple` return types. As I understand you want to use it for structural_binding, but if you introduce
```
struct ParsedConfig
{
    ExternalDataSourceConfiguration config;
    StorageSpecificArgs args;
};
```
structural binding will still work.